### ISSUE TITLE
BitVecDeque?

### ISSUE THREAD

**Issue by ckaran** [myrrlyn/bitvec#38]
I love BitVec, but really need something like a BitVecDeque so I can have an efficient bit queue (push back, pop front).  Is this possible with BitVec as it is, or would this be an enhancement?

**Comment by myrrlyn** [comment]
It is immediately available with this hack: `pop_front` is `.rotate_left(1).pop()`, and `push_front` is `.push().rotate_right(1)`. I have not yet looked at `VecDeque` in any depth, and so have not made any plans for its API to be ported.

Implementation details of `BitSlice` pointers allow any object using them to move their starting position up and down inside an element without any cost. `BitVec` does not currently expose an API to push or pop at the front, though there is no underlying restriction on doing so as long as you are willing to accept that when these events depart the base allocation address, the vector must move its buffer contents within the allocation.

I do not plan to make a channel equivalent at this time, so `BitVec` alone cannot be split across threads for concurrent add/remove operations.

I can add push/pop front methods in a future release. In the meantime, let me know if the behavior of the described workaround is sufficient.

**Comment by ckaran** [comment]
The workaround is fine for now, thank you!

**Comment by myrrlyn** [comment]
The implementation I assumed I had in play when I wrote this reply is a behavior that I scrapped in favor of a simpler, safer, but  more naïve version. `BitVec` can play much more fast-and-loose with its buffer than `BitSlice` can, so I'll write an actually-intelligent version of `push_front` and `pop_front` in the near future.

`.rotate_left(1)` moves every bit in the slice; this is not necessary in most cases. If it's a reasonably small vec, it'll still be *pretty* quick, but I can do better.
